File: hackerrank-1-week-prep-kit/day-6/jesse-and-cookies.cs
using System.CodeDom.Compiler;
using System.Collections.Generic;
using System.Collections;
using System.ComponentModel;
using System.Diagnostics.CodeAnalysis;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.Serialization;
using System.Text.RegularExpressions;
using System.Text;
using System;

class Result
{

    /*
     * Complete the 'cookies' function below.
     *
     * The function is expected to return an INTEGER.
     * The function accepts following parameters:
     *  1. INTEGER k
     *  2. INTEGER_ARRAY A
     */

    public static int cookies(int k, List<int> A)
    {
        return cookiesHelper(k, A, 0);
    }

    private static int cookiesHelper(int k, List<int> A, int count)
    {
        if (A.Count < 2) return -1;

        int minA = A.Min();
        if (minA >= k)
        {
            return count;
        }

        A.Remove(minA);
        int minB = A.Min();
        A.Remove(minB);

        A.Add(minA + 2 * minB);
        return cookiesHelper(k, A, count + 1);
    }
}

class Solution
{
    public static void Main(string[] args)
    {
        TextWriter textWriter = new StreamWriter(@System.Environment.GetEnvironmentVariable("OUTPUT_PATH"), true);

        string[] firstMultipleInput = Console.ReadLine().TrimEnd().Split(' ');

        int n = Convert.ToInt32(firstMultipleInput[0]);

        int k = Convert.ToInt32(firstMultipleInput[1]);

        List<int> A = Console.ReadLine().TrimEnd().Split(' ').ToList().Select(ATemp => Convert.ToInt32(ATemp)).ToList();

        int result = Result.cookies(k, A);

        textWriter.WriteLine(result);

        textWriter.Flush();
        textWriter.Close();
    }
}

File: hackerrank-1-week-prep-kit/old/day-6/jesse-and-cookies.py
import math
import os
import random
import re
import sys
from heapq import heapify, heappop, heappush

#
# Complete the 'cookies' function below.
#
# The function is expected to return an INTEGER.
# The function accepts following parameters:
#  1. INTEGER k
#  2. INTEGER_ARRAY A
#

def cookies(k, A):
    # Write your code here
    heapify(A)
    
    count = 0
    while A[0] < k:
        if len(A) < 2:
            return -1
            
        val = heappop(A) + 2 * heappop(A)
        heappush(A, val)
        count += 1
    
    return count
    
    
if __name__ == '__main__':
    k = 9
    A = [1,2,99,99,6,7,6]
    result = cookies(k, A)
    print(result)
